src/spatial/flatten.jl

```diff
@@ -0,0 +1,96 @@
+using DiffEqJump, DiffEqBase, LightGraphs
+
+"""
+prob.u0 must be a Matrix with prob.u0[i,j] being the number of species i at site j
+prob.p must be unscaled rates, either an array or a matrix with prob.p[i,j] being the rate of reaction j at site j
+"""
+function flatten(ma_jump::MassActionJump, prob::DiscreteProblem, spatial_system, hopping_constants; kwargs...)
+    netstoch = ma_jump.net_stoch
+    reactstoch = ma_jump.reactant_stoch
+    rx_rates = ma_jump.scaled_rates
+    tspan = prob.tspan
+    u0 = prob.u0
+    flatten(netstoch, reactstoch, rx_rates, spatial_system, u0, tspan, hopping_constants; scale_rates = false, kwargs...)
+end
+
+"""
+if hopping_constants is a matrix, assume hopping_constants[i,j] is species i, site j
+"""
+function flatten(netstoch::AbstractArray, reactstoch::AbstractArray, rx_rates::Vector, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Matrix{F}; kwargs...) where F <: Number
+    num_nodes = num_sites(spatial_system)
+    num_specs = size(u0, 1)
+    @assert size(hopping_constants) == size(u0) # hopping_constants[i,j] is species i, site j
+    hop_constants = Vector{Matrix{F}}(undef, size(hopping_constants, 2))
+    for site in 1:num_nodes
+        num_nbs = num_neighbors(spatial_system, site)
+        hop_constants[site] = reshape(repeat(hopping_constants[:,site], num_nbs), num_specs, num_nbs)
+    end
+    
+    flatten(netstoch, reactstoch, rx_rates, spatial_system, u0, tspan, hop_constants; kwargs...)
+end
+
+"""
+if reaction rates is a vector, assume reaction rates are equal across sites
+"""
+function flatten(netstoch::AbstractArray, reactstoch::AbstractArray, rx_rates::Vector, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; kwargs...) where F <: Number
+    num_nodes = num_sites(spatial_system)
+    rates = reshape(repeat(rx_rates, num_nodes), length(rx_rates), num_nodes)
+    flatten(netstoch, reactstoch, rates, spatial_system, u0, tspan, hopping_constants; kwargs...)
+end
+
+"""
+convert to Float64 if type Any
+"""
+function flatten(netstoch::AbstractArray{Any}, reactstoch::AbstractArray{Any}, rx_rates::Vector{Any}, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; kwargs...) where F <: Number
```

Comment:
Why type `Any`? We should never by working with stoichiometries that are type any.

Reply:
This happens if we do `MassActionJump([], [], [])`, which is done in, say, `diffusion.jl`.